Augmented chunk hashes based on imported css code

dionysiusmarquis opened the pull request:
### Before submitting the PR, please make sure you do the following
- [x] It's really useful if your PR references an issue where it is discussed ahead of time. In many cases, features are absent for a reason. For large changes, please create an RFC: https://github.com/sveltejs/rfcs
- [x] This message body should clearly illustrate what problems it solves.
- [ ] Ideally, include a test that fails without this PR but passes with it.

### Tests
-  [ ] Run the tests with `npm test` and lint the project with `npm run lint`

Fixes: #1660 

Currently the chunk hashes don't include "injected" css code coming from script css imports. This is because the injection is happening in bundle hooks and at this stage all depending chunk imports will already be resolved using hashes that don't "react" to css changes. This is especially crucial in caching environments.
To ensure that the chunk hashes also change as soon as imported css gets updated this PR adds a `css_hashing` plugin to the rollup compiler. The `augmentChunkHash` hook is utilised to provide a hash based on all imported css codes.

benmccann commented on src/core/create_compilers/RollupCompiler.ts:
```diff
@@ -109,6 +111,7 @@ export default class RollupCompiler {
 	errors: any[];
 	chunks: RenderedChunk[];
 	css_files: Record<string, string>;
+	css_hashes: Record<string, string>;
```
can you keep this variable contained within the `css_hashing` plugin?

benmccann commented on src/core/create_compilers/RollupCompiler.ts:
```diff
@@ -300,6 +333,14 @@ export default class RollupCompiler {
 			}
 		};
 
+		mod.plugins.push(css_hashing);
```
line 339 does `mod.plugins || (mod.plugins = [])`. it seems like this should come after that. maybe something like:
```
mod.plugins = mod.plugins || [];
mod.plugins.push(css_hashing);
mod.plugins.push(css_chunks({ injectImports: true }));
```

benmccann commented on src/core/create_compilers/RollupCompiler.ts:
```diff
@@ -154,11 +158,40 @@ export default class RollupCompiler {
+				module_ids.forEach(({ module_id, module_path }) => {
+					if (that.css_hashes[module_id]) {
+						const file = `${module_path.dir}/${module_path.name}`;
+						// ignore svelte css files created with `emitCss` enabled
```
this comment could be expanded. why ignore those files and not other css files? also, how does this check work?

benmccann commented on src/core/create_compilers/RollupCompiler.ts:
```diff
@@ -154,11 +158,40 @@ export default class RollupCompiler {
+				let hash: Hash;
+
+				const module_ids = Object.keys(chunk.modules).map(module_id => ({ module_id, module_path: path.parse(module_id) }));
+				module_ids.forEach(({ module_id, module_path }) => {
```
I think this logic might be clearer if you instead did a `filter` to find all the matching `module_id`s. Then, after that, if there are any matching IDs you can create a hash from those IDs

dionysiusmarquis replied:
I know this current trend to use chained es6 array functions. But to be honest, I never got why I should optimise script language iteration code for "readability" or "easy to understand". This is not an environment where a compiler will optimise this kind of unnecessary multi iteration for you. If you ask me I would only "allow" es6 array functions if they are really just a short variant of a "simple" iteration. I always encourage other devs I work with to first think of a solution with as few iterations as possible and, if needed, use additional variables to save or determine necessary states etc.
I have seen whole app code foundations becoming pretty laggy just because the extensive use of chained array functions, lately.
I can change it, but I wouldn't suggest it tbh (and I don't really want to participate coding this unnecessary multi iterate way, no matter how many data gets processed 😛)

benmccann replied:
I didn't mean to suggest you have to use chained es6 array functions and don't really care if you want to use `for` loops or something else. My main thought here was more about the structure of the operations. It took me awhile to understand what exactly is happening. I think if it were separated into two steps of first collecting the relevant IDs and then second building the hash from those IDs it'd be a bit easier to understand compared to building the hash as we go.

dionysiusmarquis replied:
My main point was to prefer a single iteration approach via variables (in this case `hash`) over multi iteration. I will change it, but I‘m still not a fan of using additional unnecessary iterations to make it "easier to understand".

benmccann replied:
thanks. I think the number of elements being iterated over here will be quite limited, so there should not be any measurable performance impact, which is why I prefer the readability approach

dionysiusmarquis replied:
Ok, done 😬

benmccann commented on src/core/create_compilers/RollupCompiler.ts:
```diff
@@ -154,11 +156,51 @@ export default class RollupCompiler {
+					let hash: Hash;
+					const module_files = Object.keys(chunk.modules).map(module_id => {
+						/**
+						 * Save `module_file` which represents the id without file extension. This way it's easy to determine
```
I'd probably just use `//` style comments for variables and reserve the other style for method declarations

benmccann commented on src/core/create_compilers/RollupCompiler.ts:
```diff
@@ -154,11 +156,51 @@ export default class RollupCompiler {
+		const css_hashing = () => {
+			const css_hashes: Record<string, string> = {};
+			return <Plugin>{
+				name: 'sapper-css-hashing',
```
Can this plugin be combined with the `sapper-css-injection` plugin? It seems like the hashing is really necessary because of what that plugin is doing and isn't something we'd use in a standalone fashion

dionysiusmarquis replied:
It needs to be before the css chunk plugin and is also critical for the client chunk.

dionysiusmarquis replied:
🤔 Actually this exact issue should also apply to `rollup-plugin-css-chunks` as there is also some `chunk.code` stuff happening without a corresponding `augmentChunkHash` hook. So `sapper-css-hashing` should actually fix this for the css chunk plugin as well as the sapper rollup compiler.

benmccann replied:
I was hoping to combine the `sapper-css-injection` plugin with `rollup-plugin-css-chunk`: https://github.com/domingues/rollup-plugin-css-chunks/pull/8. There's no reason that our css injection functionality couldn't be used by other Rollup projects, so I think it'd better live as a project separate from Sapper

dionysiusmarquis replied:
Thinking about it this maybe could have been solved on `rollup-plugin-css-chunks` level. I already tried to update to the newest version, but that had a major update in the mean time. Using the newest version in sapper had noticeable side effects. Fix it there might still need extra work on sapper side, I guess.

benmccann replied:
I just sent https://github.com/sveltejs/sapper/pull/1666 to update to the latest `rollup-plugin-css-chunks`. I had to bump Rollup as well to make it work

benmccann commented on src/core/create_compilers/RollupCompiler.ts:
```diff
@@ -1,3 +1,4 @@
+import { createHash, Hash } from 'crypto';
```
`lint` is complaining that `Hash` is unused

benmccann commented:
It looks like this will need to be rebased. I sent a small PR in https://github.com/sveltejs/sapper/pull/1677 and didn't realize it would conflict with this

dionysiusmarquis commented:
@benmccann This approach doesn't work anymore, since the update of `rollup-plugin-css-chunks` (https://github.com/sveltejs/sapper/commit/8a2769da8ed95c3daacfadb645946fc9822be4b1#diff-7ae45ad102eab3b6d7e7896acd08c427a9b25b346470d7bc6507b6481575d519R63).

That's because the css files won't be present in the`chunk.modules` anymore. This is most likely because of this change: https://github.com/domingues/rollup-plugin-css-chunks/commit/3632de47adea7a0447e0d53b0a608cecac1307c8#diff-dcdc3e0b3362edb8fec2a51d3fa51f8fb8af8f70247e06d9887fa934834c9122L90

Following: https://github.com/rollup/rollup/issues/3655

dionysiusmarquis commented:
@benmccann Ok, I rebased and updated the approach. Maybe you could have another look. I had to add `moduleSideEffects: 'no-treeshake'`. I was also able to simplify the `augmentChunkHash` part because I had to determine `is_svelte_css` in `transform` already.

benmccann commented:
It looks like there's a merge conflict here. This one was pretty complicated to understand and test. I wonder if it might be better to just make sure this works as intended in SvelteKit instead? SvelteKit uses Vite for CSS bundling which is a completely different set of code, so there's a good chance this issue doesn't exist there

dionysiusmarquis commented:
Everything is working as expected in SvelteKit, I'd say. Running `dev` imported css inside `<script>` will be added "unbundled" as inline css. `build` will put them in the corresponding bundled css, correctly hashed.
